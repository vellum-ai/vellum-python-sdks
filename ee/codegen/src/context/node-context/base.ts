import { VellumVariableType } from "vellum-ai/api";

import {
  VELLUM_WORKFLOW_BASE_NODES_MODULE_PATH,
  VELLUM_WORKFLOW_NODES_MODULE_PATH,
} from "src/constants";
import { WorkflowContext } from "src/context";
import { PortContext } from "src/context/port-context";
import { NodeOutputNotFoundError } from "src/generators/errors";
import { NodePort as NodePortType, WorkflowDataNode } from "src/types/vellum";
import { toValidPythonIdentifier } from "src/utils/casing";
import { getNodeLabel } from "src/utils/nodes";
import {
  doesModulePathStartWith,
  getGeneratedNodeDisplayModulePath,
  getGeneratedNodeModuleInfo,
} from "src/utils/paths";
import { hasErrorOutput } from "src/utils/typing";

export declare namespace BaseNodeContext {
  interface Args<T extends WorkflowDataNode> {
    workflowContext: WorkflowContext;
    nodeData: T;
    importOrder: readonly number[];
  }
}

export abstract class BaseNodeContext<T extends WorkflowDataNode> {
  public readonly workflowContext: WorkflowContext;
  public readonly importOrder: readonly number[];
  public readonly nodeModulePath: string[];
  public readonly nodeModuleName: string;
  public readonly nodeFileName: string;
  public readonly nodeClassName: string;
  public readonly nodeDisplayModulePath: string[];
  public readonly nodeDisplayModuleName: string;
  public readonly nodeDisplayFileName: string;
  public readonly nodeDisplayClassName: string;

  public nodeData: T;
  public abstract readonly baseNodeClassName: string;
  public abstract readonly baseNodeDisplayClassName: string;

  protected isCore = false;
  public readonly baseNodeClassModulePath: readonly string[];
  public readonly baseNodeDisplayClassModulePath: readonly string[];

  private nodeOutputNamesById: Record<string, string> | undefined;
  public readonly portContextsById: Map<string, PortContext>;
  public readonly defaultPortContext: PortContext | undefined;

  private readonly nodePortNames: Set<string> = new Set();

  constructor(args: BaseNodeContext.Args<T>) {
    this.workflowContext = args.workflowContext;
    this.nodeData = args.nodeData;
    this.importOrder = args.importOrder;

    if (args.nodeData.type === "GENERIC") {
      if (
        !doesModulePathStartWith(
          args.nodeData.base?.module,
          VELLUM_WORKFLOW_BASE_NODES_MODULE_PATH
        )
      ) {
        this.baseNodeClassModulePath = args.nodeData.base.module;
      } else {
        this.baseNodeClassModulePath = VELLUM_WORKFLOW_NODES_MODULE_PATH;
      }
    } else if (
      !doesModulePathStartWith(
        args.nodeData.base?.module,
        VELLUM_WORKFLOW_NODES_MODULE_PATH
      )
    ) {
      this.baseNodeClassModulePath =
        args.nodeData.base?.module ??
        this.workflowContext.sdkModulePathNames.DISPLAYABLE_NODES_MODULE_PATH;
    } else {
      this.baseNodeClassModulePath =
        this.workflowContext.sdkModulePathNames.DISPLAYABLE_NODES_MODULE_PATH;
    }
    this.baseNodeDisplayClassModulePath =
      this.workflowContext.sdkModulePathNames.NODE_DISPLAY_MODULE_PATH;

    const { moduleName, nodeClassName, modulePath } =
      getGeneratedNodeModuleInfo({
        workflowContext: args.workflowContext,
        nodeDefinition: args.nodeData.definition,
        nodeLabel: this.getNodeLabel(),
      });
    this.nodeModuleName = moduleName;
    this.nodeClassName = nodeClassName;
    this.nodeModulePath = modulePath;

    this.workflowContext.addUsedClassName(nodeClassName);
    this.nodeFileName = `${this.nodeModuleName}.py`;

    this.nodeDisplayModuleName = this.nodeModuleName;
    this.nodeDisplayFileName = `${this.nodeDisplayModuleName}.py`;
    this.nodeDisplayClassName = `${this.nodeClassName}Display`;
    this.nodeDisplayModulePath = getGeneratedNodeDisplayModulePath(
      args.workflowContext,
      this.nodeDisplayModuleName
    );

    const portContexts = this.nodeData.ports
      ? this.createPortContextsFromNodePortAttribute(this.nodeData.ports)
      : this.createPortContexts();
    portContexts.forEach((portContext) => {
      this.workflowContext.addPortContext(portContext);
    });

    this.portContextsById = new Map(
      portContexts.map((portContext) => [portContext.portId, portContext])
    );

    this.defaultPortContext = portContexts.find(
      (portContext) => portContext.isDefault
    );
  }

  protected abstract getNodeOutputNamesById(): Record<string, string>;
  protected abstract getNodeOutputTypesById(): Record<
    string,
    VellumVariableType
  >;
  protected abstract createPortContexts(): PortContext[];

  protected createPortContextsFromNodePortAttribute(
    nodePorts: NodePortType[]
  ): PortContext[] {
    return nodePorts.map(
      (port) =>
        new PortContext({
          workflowContext: this.workflowContext,
          nodeContext: this,
          portId: port.id,
          portName: port.name,
          isDefault: port.type === "DEFAULT",
        })
    );
  }

  public getNodeLabel(): string {
    return getNodeLabel(this.nodeData);
  }

  public getNodeOutputNameById(outputId: string): string | undefined {
    // Lazily load node output names
    if (!this.nodeOutputNamesById) {
      this.nodeOutputNamesById = this.getNodeOutputNamesById();
    }

    const nodeOutputName = this.nodeOutputNamesById[outputId];
    if (!nodeOutputName) {
      // This handles an edge case where the node references a non-existent subworkflow deployment
      if (
        this.nodeData.type === "SUBWORKFLOW" &&
        this.nodeData.data.variant === "DEPLOYMENT"
      ) {
        this.workflowContext.addError(
          new NodeOutputNotFoundError(
            `Could not find Subworkflow Deployment Output with id ${outputId}`,
            "WARNING"
          )
        );
        return;
      }

      this.workflowContext.addError(
        new NodeOutputNotFoundError(
          `Failed to find ${this.nodeClassName} Output with id '${outputId}'`,
          "WARNING"
        )
      );
      return;
    }

    return toValidPythonIdentifier(nodeOutputName, "output");
  }

  public getNodeOutputTypeById(
    outputId: string
  ): VellumVariableType | undefined {
    const nodeOutputTypesById = this.getNodeOutputTypesById();

    return nodeOutputTypesById[outputId];
  }

  public isPortNameUsed(portName: string): boolean {
    return this.nodePortNames.has(portName);
  }

  public addUsedPortName(portName: string): void {
    this.nodePortNames.add(portName);
  }

  /**
   * Extend this class to perform any additional property generation asynchronously
   * after the node context has been created.
   */
  public buildProperties(): Promise<void> {
    return Promise.resolve();
  }

  public isImportedBefore(
    nodeContext: BaseNodeContext<WorkflowDataNode>
  ): boolean {
    for (
      let i = 0;
      i < Math.max(this.importOrder.length, nodeContext.importOrder.length);
      i++
    ) {
      const thisImportOrder = this.importOrder[i];
      const otherImportOrder = nodeContext.importOrder[i];
      if (thisImportOrder === undefined && otherImportOrder === undefined) {
        return false;
      }

      if (thisImportOrder === undefined) {
        return true;
      }

      if (otherImportOrder === undefined) {
        return false;
      }

      if (thisImportOrder < otherImportOrder) {
        return true;
      }

      if (thisImportOrder > otherImportOrder) {
        return false;
      }
    }
    return false;
  }

  protected getErrorOutputId(): string | undefined {
    const TRY_NODE_NAME = "TryNode";
    const tryNodes = this.nodeData.adornments?.filter((adornment) => {
      return adornment.base.name === TRY_NODE_NAME;
    });

    const tryNode = tryNodes?.at(-1);

    let fallbackErrorOutputId: string | undefined = undefined;
    if (hasErrorOutput(this.nodeData)) {
      fallbackErrorOutputId = this.nodeData.data.errorOutputId;
    }

    return tryNode ? tryNode.id : fallbackErrorOutputId;
  }
}
