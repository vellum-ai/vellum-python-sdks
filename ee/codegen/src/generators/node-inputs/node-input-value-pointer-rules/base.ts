import { AstNode } from "@fern-api/python-ast/core/AstNode";
import { Writer } from "@fern-api/python-ast/core/Writer";

import { WorkflowContext } from "src/context";
import { BaseNodeContext } from "src/context/node-context/base";
import {
  IterableConfig,
  NodeInputValuePointerRule as NodeInputValuePointerRuleType,
  WorkflowDataNode,
} from "src/types/vellum";

export declare namespace BaseNodeInputValuePointerRule {
  export interface Args<T extends NodeInputValuePointerRuleType> {
    nodeContext: BaseNodeContext<WorkflowDataNode>;
    nodeInputValuePointerRule: T;
    iterableConfig?: IterableConfig;
  }
}

export abstract class BaseNodeInputValuePointerRule<
  T extends NodeInputValuePointerRuleType
> extends AstNode {
  private readonly nodeContext: BaseNodeContext<WorkflowDataNode>;
  public readonly workflowContext: WorkflowContext;
  public readonly nodeInputValuePointerRule: T;
  public readonly iterableConfig?: IterableConfig;
  private astNode: AstNode | undefined;

  constructor({
    nodeContext,
    nodeInputValuePointerRule,
    iterableConfig,
  }: BaseNodeInputValuePointerRule.Args<T>) {
    super();
    this.nodeContext = nodeContext;
    this.workflowContext = nodeContext.workflowContext;
    this.iterableConfig = iterableConfig;
    this.nodeInputValuePointerRule = nodeInputValuePointerRule;

    this.astNode = this.getAstNode();
    if (this.astNode) {
      this.inheritReferences(this.astNode);
    }
  }

  public getReferencedNodeContext():
    | BaseNodeContext<WorkflowDataNode>
    | undefined {
    return undefined;
  }

  abstract getAstNode(): AstNode | undefined;

  public write(writer: Writer): void {
    if (this.astNode) {
      this.astNode.write(writer);
    }
  }
}
