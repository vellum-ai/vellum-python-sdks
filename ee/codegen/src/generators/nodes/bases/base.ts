import { python } from "@fern-api/python-ast";
import { MethodArgument } from "@fern-api/python-ast/MethodArgument";
import { AstNode } from "@fern-api/python-ast/core/AstNode";

import * as codegen from "src/codegen";
import { PORTS_CLASS_NAME } from "src/constants";
import { WorkflowContext } from "src/context";
import { BaseNodeContext } from "src/context/node-context/base";
import {
  BaseCodegenError,
  NodeAttributeGenerationError,
} from "src/generators/errors";
import { NodeDisplayData } from "src/generators/node-display-data";
import { NodeInput } from "src/generators/node-inputs/node-input";
import { NodePorts } from "src/generators/node-port";
import { NODE_DEFAULT_ATTRIBUTES } from "src/generators/nodes/constants";
import { UuidOrString } from "src/generators/uuid-or-string";
import { WorkflowValueDescriptor } from "src/generators/workflow-value-descriptor";
import { WorkflowProjectGenerator } from "src/project";
import {
  AdornmentNode,
  NodeDisplayComment,
  NodeDisplayData as NodeDisplayDataType,
  WorkflowDataNode,
} from "src/types/vellum";
import { isNilOrEmpty } from "src/utils/typing";

export declare namespace BaseNode {
  interface Args<T extends WorkflowDataNode, V extends BaseNodeContext<T>> {
    workflowContext: WorkflowContext;
    nodeContext: V;
  }
}

export abstract class BaseNode<
  T extends WorkflowDataNode,
  V extends BaseNodeContext<T>
> {
  public readonly workflowContext: WorkflowContext;
  public readonly nodeData: T;
  public readonly nodeContext: V;

  protected readonly nodeInputsByKey: Map<string, NodeInput>;
  protected readonly nodeInputsById: Map<string, NodeInput>;
  protected readonly nodeAttributeNameByNodeInputId: Map<string, string>;

  private readonly errorOutputId: string | undefined;

  constructor({ workflowContext, nodeContext }: BaseNode.Args<T, V>) {
    this.workflowContext = workflowContext;
    this.nodeContext = nodeContext;
    this.nodeData = nodeContext.nodeData;

    [
      this.nodeInputsByKey,
      this.nodeInputsById,
      this.nodeAttributeNameByNodeInputId,
    ] = this.generateNodeInputs();
    this.errorOutputId = this.getErrorOutputId();
  }

  public abstract persist(): Promise<void>;

  protected abstract getNodeClassBodyStatements(): AstNode[];

  protected abstract getNodeDisplayClassBodyStatements(): AstNode[];

  // Override to specify a custom output display
  protected abstract getOutputDisplay(): python.Field | undefined;

  // If the node supports the Reject on Error toggle, then implement this to return
  // the error_output_id from this.nodeData. If returned, a @TryNode decorator will be
  // added to the node class.
  protected abstract getErrorOutputId(): string | undefined;

  // Override if the node implementation's base class needs to include generic types
  protected getNodeBaseGenericTypes(): AstNode[] | undefined {
    return undefined;
  }

  protected getNodeBaseClass(): python.Reference {
    const baseNodeClassNameAlias =
      this.nodeContext.baseNodeClassName === this.nodeContext.nodeClassName
        ? `Base${this.nodeContext.baseNodeClassName}`
        : undefined;

    const baseNodeGenericTypes = this.getNodeBaseGenericTypes();

    return python.reference({
      name: this.nodeContext.baseNodeClassName,
      modulePath: this.nodeContext.baseNodeClassModulePath,
      genericTypes: baseNodeGenericTypes,
      alias: baseNodeClassNameAlias,
    });
  }

  protected getNodeDisplayBaseClass(): python.Reference {
    return python.reference({
      name: this.nodeContext.baseNodeDisplayClassName,
      modulePath: this.nodeContext.baseNodeDisplayClassModulePath,
      genericTypes: [
        python.reference({
          name: this.nodeContext.nodeClassName,
          modulePath: this.nodeContext.nodeModulePath,
        }),
      ],
    });
  }

  /* Override if the node implementation needs to generate nested workflows */
  protected generateNestedWorkflowContexts(): Map<string, WorkflowContext> {
    return new Map();
  }

  /* Override if the node implementation needs to generate nested workflows */
  protected generateNestedProjectsByName(): Map<
    string,
    WorkflowProjectGenerator
  > {
    return new Map();
  }

  protected findNodeInputByName(name: string): NodeInput | undefined {
    return this.nodeInputsByKey.get(name);
  }

  protected getNodeInputByName(name: string): NodeInput | undefined {
    const nodeInput = this.findNodeInputByName(name);
    if (!nodeInput) {
      this.workflowContext.addError(
        new NodeAttributeGenerationError(
          `No input found named "${name}"`,
          "WARNING"
        )
      );
    }

    return nodeInput;
  }

  public getNodeClassName() {
    return this.nodeContext.nodeClassName;
  }

  public getNodeModulePath() {
    return this.nodeContext.nodeModulePath;
  }

  public getNodeDisplayClassName() {
    return this.nodeContext.nodeDisplayClassName;
  }

  public getNodeDisplayModulePath() {
    return this.nodeContext.nodeDisplayModulePath;
  }

  protected getNodeAttributeNameByNodeInputKey(nodeInputKey: string): string {
    /**
     * This method drives how we map the key to a legacy node input to the new
     * node attribute name in the SDK.
     *
     * By default, we just pass through. However, legacy nodes can extend this
     * method to customize their specific mappings.
     */
    return nodeInputKey;
  }

  private generateNodeInputs(): [
    Map<string, NodeInput>,
    Map<string, NodeInput>,
    Map<string, string>
  ] {
    const nodeInputsByKey = new Map<string, NodeInput>();
    const nodeInputsById = new Map<string, NodeInput>();
    const nodeAttributeNameByNodeInputId = new Map<string, string>();

    if (!("inputs" in this.nodeData)) {
      return [nodeInputsByKey, nodeInputsById, nodeAttributeNameByNodeInputId];
    }

    this.nodeData.inputs.forEach((nodeInputData) => {
      try {
        const nodeInput = codegen.nodeInput({
          nodeContext: this.nodeContext,
          nodeInputData,
        });

        if (!isNilOrEmpty(nodeInput.nodeInputValuePointer.rules)) {
          nodeInputsByKey.set(nodeInputData.key, nodeInput);
          nodeInputsById.set(nodeInputData.id, nodeInput);
          const nodeAttributeName = this.getNodeAttributeNameByNodeInputKey(
            nodeInputData.key
          );
          nodeAttributeNameByNodeInputId.set(
            nodeInputData.id,
            nodeAttributeName
          );
        }
      } catch (error) {
        if (error instanceof BaseCodegenError) {
          const nodeAttributeGenerationError = new NodeAttributeGenerationError(
            `Failed to generate attribute '${this.nodeContext.nodeClassName}.inputs.${nodeInputData.key}': ${error.message}`,
            "WARNING"
          );
          this.workflowContext.addError(nodeAttributeGenerationError);
        } else {
          throw error;
        }
      }
    });

    return [nodeInputsByKey, nodeInputsById, nodeAttributeNameByNodeInputId];
  }

  protected getPortDisplay(): python.Field | undefined {
    if (
      !("data" in this.nodeData) ||
      !("sourceHandleId" in this.nodeData.data)
    ) {
      return;
    }

    return python.field({
      name: "port_displays",
      initializer: python.TypeInstantiation.dict([
        {
          key: python.reference({
            name: this.nodeContext.nodeClassName,
            modulePath: this.nodeContext.nodeModulePath,
            attribute: [PORTS_CLASS_NAME, "default"],
          }),
          value: python.instantiateClass({
            classReference: python.reference({
              name: "PortDisplayOverrides",
              modulePath:
                this.workflowContext.sdkModulePathNames
                  .NODE_DISPLAY_TYPES_MODULE_PATH,
            }),
            arguments_: [
              python.methodArgument({
                name: "id",
                value: python.TypeInstantiation.uuid(
                  this.nodeData.data.sourceHandleId
                ),
              }),
            ],
          }),
        },
      ]),
    });
  }

  private generateNodeDisplayDataWithoutComment() {
    const nodeDisplayData: NodeDisplayDataType | undefined =
      this.nodeData.displayData;
    if (
      nodeDisplayData &&
      nodeDisplayData.comment &&
      nodeDisplayData.comment.expanded
    ) {
      const nodeDisplayDataWithoutComment: NodeDisplayDataType = {
        position: nodeDisplayData.position,
        width: nodeDisplayData.width,
        height: nodeDisplayData.height,
        comment: {
          expanded: nodeDisplayData.comment.expanded,
        },
      };
      return new NodeDisplayData({
        workflowContext: this.workflowContext,
        nodeDisplayData: nodeDisplayDataWithoutComment,
      });
    }
    return new NodeDisplayData({
      workflowContext: this.workflowContext,
      nodeDisplayData: this.nodeData.displayData,
    });
  }

  private getDisplayData(): python.Field {
    return python.field({
      name: "display_data",
      initializer: this.generateNodeDisplayDataWithoutComment(),
    });
  }

  protected getAdornments(): AdornmentNode[] {
    if (this.nodeData.adornments && this.nodeData.adornments.length > 0) {
      // TODO: Add validation for adornments
      return this.nodeData.adornments;
    }

    return [];
  }

  protected getNodeDecorators(): python.Decorator[] | undefined {
    const decorators: python.Decorator[] = [];
    const errorOutputId = this.getErrorOutputId();

    if (errorOutputId) {
      decorators.push(
        python.decorator({
          callable: python.invokeMethod({
            methodReference: python.reference({
              name: "TryNode",
              attribute: ["wrap"],
              modulePath:
                this.workflowContext.sdkModulePathNames.CORE_NODES_MODULE_PATH,
            }),
            arguments_: [],
          }),
        })
      );
    }

    const adornments = this.getAdornments();

    for (const adornment of adornments) {
      // TODO: remove this check when we remove errorOutputId
      if (errorOutputId && adornment.base.name === "TryNode") {
        continue;
      }

      // Filter out attributes that match their default values
      const filteredAttributes = adornment.attributes.filter((attr) =>
        this.filterAttribute(adornment.base.name, attr.name, attr.value)
      );

      if (adornment.base) {
        decorators.push(
          python.decorator({
            callable: python.invokeMethod({
              methodReference: python.reference({
                name: adornment.base.name,
                attribute: ["wrap"],
                modulePath: adornment.base.module,
              }),
              arguments_: filteredAttributes.map((attr) =>
                python.methodArgument({
                  name: attr.name,
                  value: new WorkflowValueDescriptor({
                    workflowValueDescriptor: attr.value,
                    nodeContext: this.nodeContext,
                    workflowContext: this.workflowContext,
                    iterableConfig: { endWithComma: false },
                  }),
                })
              ),
            }),
          })
        );
      }
    }

    return decorators.length > 0 ? decorators : undefined;
  }

  protected getNodePorts(): AstNode | undefined {
    if (this.nodeData.ports) {
      return new NodePorts({
        nodePorts: this.nodeData.ports,
        nodeContext: this.nodeContext,
        workflowContext: this.workflowContext,
      });
    } else {
      return undefined;
    }
  }

  public generateNodeClass(): python.Class {
    const nodeContext = this.nodeContext;

    let nodeBaseClass: python.Reference = this.getNodeBaseClass();
    if (nodeBaseClass.name === nodeContext.nodeClassName) {
      nodeBaseClass = python.reference({
        name: nodeBaseClass.name,
        modulePath: nodeBaseClass.modulePath,
        genericTypes: nodeBaseClass.genericTypes,
        alias: `Base${nodeBaseClass.name}`,
        attribute: nodeBaseClass.attribute,
      });
    }

    const nodeClass = python.class_({
      name: nodeContext.nodeClassName,
      extends_: [nodeBaseClass],
      docs: this.generateNodeComment(),
      decorators: this.getNodeDecorators(),
    });

    try {
      this.getNodeClassBodyStatements().forEach((statement) =>
        nodeClass.add(statement)
      );
      const nodePorts = this.getNodePorts();

      if (nodePorts) {
        nodeClass.add(nodePorts);
      }
    } catch (error) {
      if (error instanceof BaseCodegenError) {
        this.workflowContext.addError(error);
      } else {
        throw error;
      }
    }

    return nodeClass;
  }

  public generateNodeDisplayClasses(): python.Class[] {
    const nodeContext = this.nodeContext;
    const decorators: python.Decorator[] = [];
    const errorOutputId = this.getErrorOutputId();

    if (errorOutputId) {
      decorators.push(
        python.decorator({
          callable: python.invokeMethod({
            methodReference: python.reference({
              name: "BaseTryNodeDisplay",
              attribute: ["wrap"],
              modulePath:
                this.workflowContext.sdkModulePathNames
                  .NODE_DISPLAY_MODULE_PATH,
            }),
            arguments_: [
              // TODO: Node id and error output id are the same here for legacy reasons. Will no longer be the
              // case once we define output transformations for node adornments and could remove error output id.
              // https://linear.app/vellum/issue/APO-213/define-output-transformations-for-node-adornments
              new MethodArgument({
                name: "node_id",
                value: python.TypeInstantiation.uuid(errorOutputId),
              }),
              new MethodArgument({
                name: "error_output_id",
                value: python.TypeInstantiation.uuid(errorOutputId),
              }),
            ],
          }),
        })
      );
    }
    const adornments = this.getAdornments();

    for (const adornment of adornments) {
      // TODO: remove this check when we remove errorOutputId
      if (errorOutputId && adornment.base.name === "TryNode") {
        continue;
      }

      if (adornment.base) {
        decorators.push(
          python.decorator({
            callable: python.invokeMethod({
              methodReference: python.reference({
                name: `Base${adornment.base.name}Display`,
                attribute: ["wrap"],
                modulePath:
                  this.workflowContext.sdkModulePathNames
                    .NODE_DISPLAY_MODULE_PATH,
              }),
              // TODO: When we define output transformations, that's what we'd use here. eg, `error_output_id`.
              // https://linear.app/vellum/issue/APO-213/define-output-transformations-for-node-adornments
              arguments_: [
                new MethodArgument({
                  name: "node_id",
                  value: python.TypeInstantiation.uuid(adornment.id),
                }),
              ],
            }),
          })
        );
      }
    }

    const nodeClass = python.class_({
      name: nodeContext.nodeDisplayClassName,
      extends_: [this.getNodeDisplayBaseClass()],
      decorators: decorators.length > 0 ? decorators : undefined,
    });

    this.getNodeDisplayClassBodyStatements().forEach((statement) =>
      nodeClass.add(statement)
    );

    const nodeInputIdsByNameField = python.field({
      name: "node_input_ids_by_name",
      initializer: python.TypeInstantiation.dict(
        Array.from(this.nodeInputsByKey).map<{
          key: AstNode;
          value: AstNode;
        }>(([key, nodeInput]) => {
          const nodeAttributeName = this.nodeAttributeNameByNodeInputId.get(
            nodeInput.nodeInputData.id
          );
          return {
            key: python.TypeInstantiation.str(nodeAttributeName ?? key),
            value: new UuidOrString(nodeInput.nodeInputData.id),
          };
        })
      ),
    });
    nodeClass.add(nodeInputIdsByNameField);

    try {
      const outputDisplay = this.getOutputDisplay();
      if (outputDisplay) {
        nodeClass.add(outputDisplay);
      }
    } catch (error) {
      if (error instanceof BaseCodegenError) {
        this.workflowContext.addError(error);
      } else {
        throw error;
      }
    }

    const portDisplay = this.getPortDisplay();
    if (portDisplay) {
      nodeClass.add(portDisplay);
    }

    nodeClass.add(this.getDisplayData());

    return [nodeClass];
  }

  private getNodeComment(): NodeDisplayComment | undefined {
    return this.nodeData.displayData && "comment" in this.nodeData.displayData
      ? this.nodeData.displayData.comment
      : undefined;
  }

  private generateNodeComment(): string | undefined {
    const nodeComment = this.getNodeComment();

    if (nodeComment && nodeComment.value) {
      let comment = nodeComment.value;
      // if comment start with ", add a \ in front of it
      if (comment.startsWith('"')) {
        comment = `\\${comment}`;
      }
      // if comment end with ", add a \ in front of it
      if (comment.endsWith('"')) {
        comment = `${comment.slice(0, -1)}\\"`;
      }
      return comment;
    }
    return undefined;
  }

  private filterAttribute(
    nodeName: string,
    attributeName: string,
    attributeValue: unknown
  ): boolean {
    const nodeConfig = NODE_DEFAULT_ATTRIBUTES[nodeName];
    if (!nodeConfig) {
      return true; // Include if no config exists for this node
    }

    const attrConfig = nodeConfig[attributeName];
    if (!attrConfig) {
      return true; // Include if no config exists for this attribute
    }

    return attrConfig.defaultValue !== attributeValue; // Include if value differs from default
  }
}
