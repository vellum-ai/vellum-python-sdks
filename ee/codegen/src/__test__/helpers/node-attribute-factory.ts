import { NodeAttribute } from "src/types/vellum";

export const nodeAttributeFactory = (
  id: string,
  name: string,
  attributes: NodeAttribute[]
): NodeAttribute => ({
  id,
  name,
  value: {
    type: "CONSTANT_VALUE",
    value: {
      type: "JSON",
      value: attributes,
    },
  },
});
