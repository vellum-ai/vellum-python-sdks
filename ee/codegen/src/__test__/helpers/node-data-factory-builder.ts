import {
  AdornmentNode,
  CodeResourceDefinition,
  NodeAttribute,
  NodeOutput,
  NodePort,
  NodeTrigger,
  WorkflowDataNode as WorkflowDataNodeType,
} from "src/types/vellum";

export class NodeDataFactoryBuilder<T extends WorkflowDataNodeType> {
  public nodeData: T;

  constructor(nodeData: T) {
    this.nodeData = nodeData;
  }

  withPorts(ports: NodePort[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      ports: ports,
    };
    return this;
  }

  withAttributes(attributes: NodeAttribute[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      attributes,
    };
    return this;
  }

  withAdornments(adornments: AdornmentNode[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      adornments,
    };
    return this;
  }

  withOutputs(outputs: NodeOutput[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      outputs,
    };
    return this;
  }

  withBase(base: CodeResourceDefinition): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      base,
    };
    return this;
  }

  withTrigger(trigger: NodeTrigger): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      trigger,
    };
    return this;
  }

  build(): T {
    return this.nodeData;
  }
}
