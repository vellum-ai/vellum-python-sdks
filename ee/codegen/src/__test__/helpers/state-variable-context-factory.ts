import { VellumVariable } from "vellum-ai/api/types";

import { WorkflowContext } from "src/context";
import { StateVariableContext } from "src/context/state-variable-context";

export function stateVariableContextFactory({
  stateVariableData,
  workflowContext,
}: {
  stateVariableData: VellumVariable;
  workflowContext: WorkflowContext;
}): StateVariableContext {
  return new StateVariableContext({
    stateVariableData,
    workflowContext,
  });
}
