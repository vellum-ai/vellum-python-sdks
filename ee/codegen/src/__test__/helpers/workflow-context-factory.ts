import { entrypointNodeDataFactory } from "./node-data-factories";

import { WorkflowContext } from "src/context";

export function workflowContextFactory({
  absolutePathToOutputDirectory,
  moduleName,
  workflowClassName,
  workflowRawData,
  strict = true,
  disableFormatting = true,
}: Partial<WorkflowContext.Args> = {}): WorkflowContext {
  return new WorkflowContext({
    absolutePathToOutputDirectory:
      absolutePathToOutputDirectory || "./src/__tests__/",
    moduleName: moduleName || "code",
    workflowClassName: workflowClassName || "TestWorkflow",
    vellumApiKey: "<TEST_API_KEY>",
    workflowRawData: workflowRawData || {
      nodes: [entrypointNodeDataFactory()],
      edges: [],
    },
    strict,
    disableFormatting,
  });
}
