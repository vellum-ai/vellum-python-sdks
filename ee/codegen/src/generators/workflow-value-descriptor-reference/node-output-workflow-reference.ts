import { python } from "@fern-api/python-ast";
import { AstNode } from "@fern-api/python-ast/core/AstNode";

import { OUTPUTS_CLASS_NAME } from "src/constants";
import { BaseNodeInputWorkflowReference } from "src/generators/workflow-value-descriptor-reference/BaseNodeInputWorkflowReference";
import { NodeOutputWorkflowReference as NodeOutputWorkflowReferenceType } from "src/types/vellum";

export class NodeOutputWorkflowReference extends BaseNodeInputWorkflowReference<NodeOutputWorkflowReferenceType> {
  getAstNode(): AstNode | undefined {
    const nodeOutputPointer = this.nodeInputWorkflowReferencePointer;

    const nodeContext = this.workflowContext.getNodeContext(
      nodeOutputPointer.nodeId
    );

    const nodeOutputName = nodeContext.getNodeOutputNameById(
      nodeOutputPointer.nodeOutputId
    );

    if (nodeOutputName) {
      return python.reference({
        name: nodeContext.nodeClassName,
        modulePath: nodeContext.nodeModulePath,
        attribute: [OUTPUTS_CLASS_NAME, nodeOutputName],
      });
    }
    return undefined;
  }
}
