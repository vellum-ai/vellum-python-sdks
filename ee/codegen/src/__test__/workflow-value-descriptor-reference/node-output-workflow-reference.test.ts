import { Writer } from "@fern-api/python-ast/core/Writer";

import { workflowContextFactory } from "src/__test__/helpers";
import { WorkflowContext } from "src/context";
import { BaseNodeContext } from "src/context/node-context/base";
import { NodeOutputWorkflowReference } from "src/generators/workflow-value-descriptor-reference/node-output-workflow-reference";
import {
  WorkflowDataNode,
  WorkflowValueDescriptorReference,
} from "src/types/vellum";

describe("NodeOutputWorkflowReferencePointer", () => {
  let workflowContext: WorkflowContext;

  beforeEach(() => {
    workflowContext = workflowContextFactory();
    vi.spyOn(workflowContext, "getNodeContext").mockReturnValue({
      nodeClassName: "TestNode",
      path: ["nodes", "test-node-path"],
      getNodeOutputNameById: vi.fn().mockReturnValue("my_output"),
    } as unknown as BaseNodeContext<WorkflowDataNode>);
  });

  it("should generate correct AST for node output reference", async () => {
    const nodeOutputReference: WorkflowValueDescriptorReference = {
      type: "NODE_OUTPUT",
      nodeId: "test-node",
      nodeOutputId: "test-output",
    };

    const pointer = new NodeOutputWorkflowReference({
      workflowContext,
      nodeInputWorkflowReferencePointer: nodeOutputReference,
    });

    const writer = new Writer();
    pointer.write(writer);

    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });
});
