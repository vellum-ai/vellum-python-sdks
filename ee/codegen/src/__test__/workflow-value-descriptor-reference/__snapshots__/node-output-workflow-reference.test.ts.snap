// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NodeOutputWorkflowReferencePointer > should generate correct AST for node output reference 1`] = `
"TestNode.Outputs.my_output
"
`;
